packages/react/src/components/UIShell/SideNavSwitcher.tsx

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright IBM Corp. 2016, 2023
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { ChevronDown } from '@carbon/icons-react';
+import cx from 'classnames';
+import PropTypes from 'prop-types';
+import React, { ChangeEventHandler, forwardRef, useCallback } from 'react';
+import { useId } from '../../internal/useId';
+import { usePrefix } from '../../internal/usePrefix';
+
+interface SideNavSwitcherProps {
+  /**
+   * Provide an optional class to be applied to the containing node
+   */
+  className?: string;
+
+  /**
+   * Provide the label for the switcher. This will be the first visible option
+   * when someone views this control
+   */
+  labelText: string;
+
+  /**
+   * Provide a callback function that is called whenever the switcher value is
+   * updated
+   */
+  onChange?: ChangeEventHandler<HTMLSelectElement>;
+
+  /**
+   * Provide an array of options to be rendered in the switcher as an
+   * `<option>`. The text value will be what is displayed to the user and is set
+   * as the `value` prop for each `<option>`.
+   */
+  options: string[];
+}
+
+const SideNavSwitcher = forwardRef<HTMLSelectElement, SideNavSwitcherProps>(
+  function SideNavSwitcher(props, ref) {
+    const id = useId('side-nav-switcher');
+    const prefix = usePrefix();
+    const { className: customClassName, labelText, onChange, options } = props;
+    const className = cx(`${prefix}--side-nav__switcher`, customClassName);
+
+    const handleOnChange = useCallback<ChangeEventHandler<HTMLSelectElement>>(
+      (event) => {
+        if (onChange) {
+          onChange(event);
+        }
+      },
+      [onChange]
+    );
```

Comment:
Unless I'm mistaken, if the reference to the `onChange` passed into this component isn't stable, it's still going to re-render because `onChange` is in the dependency array and will be a new reference. The memoization needs to happen outside of the component.

The renders here are super cheap, too. I'm not sure if the useCallback is needed. What do you think?

Reply:
You're right, it will still change the dep, good catch. I'll back it out now

Reply:
reverted